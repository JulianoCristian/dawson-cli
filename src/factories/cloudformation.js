export const SAFE_STACK_POLICY = {
  // https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/protect-stack-resources.html
  // DynamoDB tables & S3 Buckets shall not be deleted
  Statement: [
    {
      Effect: 'Deny',
      Principal: '*',
      Action: ['Update:Replace', 'Update:Delete'],
      Resource: '*',
      Condition: {
        StringEquals: {
          ResourceType: [
            'AWS::DynamoDB::Table',
            'AWS::ApiGateway::RestApi',
            'AWS::CloudFront::Distribution',
            'AWS::S3::Bucket'
          ]
        }
      }
    },
    { Effect: 'Allow', Principal: '*', Action: 'Update:*', Resource: '*' }
  ]
};

export const UNSAFE_STACK_POLICY = {
  Statement: [
    { Effect: 'Allow', Action: 'Update:*', Principal: '*', Resource: '*' }
  ]
};

export function templateStackName ({ appName, stage }) {
  const stageUCFirst = stage ? stage[0].toUpperCase() + stage.substring(1) : '';
  return `${appName}${stageUCFirst}`;
}

export function buildCreateStackParams (
  { stackName, cfTemplateJSON, templateURL, inline = false }
) {
  if (inline === false && !templateURL) {
    throw new Error(`Internal error: templateUrl is required when calling buildCreateStackParams with inline === false`);
  }
  if (inline === true && !cfTemplateJSON) {
    throw new Error(`Internal error: cfTemplateJSON is required when calling buildCreateStackParams with inline === true`);
  }
  const templateSource = inline
    ? { TemplateBody: cfTemplateJSON }
    : { TemplateURL: templateURL };
  var params = {
    StackName: stackName,
    Capabilities: ['CAPABILITY_IAM', 'CAPABILITY_NAMED_IAM'],
    Tags: [{ Key: 'createdBy', Value: 'dawson' }],
    ...templateSource,
    StackPolicyBody: JSON.stringify(SAFE_STACK_POLICY),
    OnFailure: 'DO_NOTHING' // deleted when updating
  };
  return params;
}
