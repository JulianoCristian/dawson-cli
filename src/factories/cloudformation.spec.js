import test from 'ava';
import { buildCreateStackParams } from './cloudformation';

test('creates cloudformation updateStack params with an inline source', t => {
  t.snapshot(
    buildCreateStackParams({
      stackName: 'bar',
      cfTemplateJSON: JSON.stringify({ Resources: {}, Outputs: {} }),
      inline: true
    })
  );
});

test('creates cloudformation updateStack params with source from S3', t => {
  t.snapshot(
    buildCreateStackParams({ stackName: 'bar', templateURL: 'https://s3....' })
  );
});
