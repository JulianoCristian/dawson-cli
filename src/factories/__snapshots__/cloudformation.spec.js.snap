module.exports[`creates cloudformation updateStack params with an inline source`] = {
  StackName: `bar`,
  Capabilities: [
    `CAPABILITY_IAM`,
    `CAPABILITY_NAMED_IAM`
  ],
  Tags: [
    {
      Key: `createdBy`,
      Value: `dawson`
    }
  ],
  TemplateBody: `{"Resources":{},"Outputs":{}}`,
  StackPolicyBody: `{"Statement":[{"Effect":"Deny","Principal":"*","Action":["Update:Replace","Update:Delete"],"Resource":"*","Condition":{"StringEquals":{"ResourceType":["AWS::DynamoDB::Table","AWS::ApiGateway::RestApi","AWS::CloudFront::Distribution","AWS::S3::Bucket"]}}},{"Effect":"Allow","Principal":"*","Action":"Update:*","Resource":"*"}]}`,
  OnFailure: `DO_NOTHING`
}

module.exports[`creates cloudformation updateStack params with source from S3`] = {
  StackName: `bar`,
  Capabilities: [
    `CAPABILITY_IAM`,
    `CAPABILITY_NAMED_IAM`
  ],
  Tags: [
    {
      Key: `createdBy`,
      Value: `dawson`
    }
  ],
  TemplateURL: `https://s3....`,
  StackPolicyBody: `{"Statement":[{"Effect":"Deny","Principal":"*","Action":["Update:Replace","Update:Delete"],"Resource":"*","Condition":{"StringEquals":{"ResourceType":["AWS::DynamoDB::Table","AWS::ApiGateway::RestApi","AWS::CloudFront::Distribution","AWS::S3::Bucket"]}}},{"Effect":"Allow","Principal":"*","Action":"Update:*","Resource":"*"}]}`,
  OnFailure: `DO_NOTHING`
}